File: LeetCode/Hard/10 RegularExpressionMatching/10RegularExpressionMatching.py
def is_match(s: str, p: str) -> bool:
    m, n = len(s), len(p)

    # Lista de listas en vez de vector<vector<bool>>. Se usa una comprensión de listas anidada para crear cada fila de forma independiente (evita el bug clásico de Python donde
    # [[False]*n]*m crea filas que en realidad son la MISMA lista repetida por referencia).
    dp = [[False] * (n + 1) for _ in range(m + 1)]
    dp[0][0] = True

    # Fila 0: patrón contra cadena vacía.
    for j in range(1, n + 1):
        if p[j - 1] == '*':
            dp[0][j] = dp[0][j - 2]

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if p[j - 1] == '.' or p[j - 1] == s[i - 1]:
                # coincidencia directa, se arrastra el resultado anterior
                dp[i][j] = dp[i - 1][j - 1]
            elif p[j - 1] == '*':
                usar_cero = dp[i][j - 2]       # usar X* cero veces
                usar_una_mas = False           # usar X* una vez mas
                caracter_anterior = p[j - 2]
                if caracter_anterior == '.' or caracter_anterior == s[i - 1]:
                    usar_una_mas = dp[i - 1][j]
                dp[i][j] = usar_cero or usar_una_mas
            else:
                dp[i][j] = False

    return dp[m][n]


def main():
    s1, p1 = "aa", "a"
    print(f'Salida del problema 10 ("{s1}", "{p1}"): {is_match(s1, p1)}')

if __name__ == "__main__":
    main()

File: LeetCode/Hard/10 RegularExpressionMatching/10RegularExpressionMatching.cpp
#include <iostream>
#include <string>
#include <vector>

class Solution {
public:
    bool isMatch(std::string s, std::string p) {        
        int m = s.length(), n = p.length();
        std::vector<std::vector<bool>> dp (m+1, std::vector<bool>(n+1, false));
        dp[0][0] = true;
        
        // Fila 0: patrón contra cadena vacía. Solo puede ser true, si el patrón es una secuencia de "X*" (ej: "a*b*c*")
        for(int j=1; j<=n; j++) {
            if(p[j-1] == '*') {
                dp[0][j] = dp[0][j-2];
            }
        }

        for(int i=1; i<=m; i++) {
            for(int j=1; j<=n; j++) {
                if(p[j-1] == '.' || p[j-1] == s[i-1]) {
                    dp[i][j] = dp[i-1][j-1]; // coincidencia directa por lo que arrastro el resultado anterior;
                } else if(p[j-1] == '*') {
                    bool usarCero = dp[i][j-2]; // usar X* cero veces
                    bool usarUnaMas = false; // usar X* una vez mas 
                    char characterAnterior = p[j-2];
                    if(characterAnterior == '.' || characterAnterior == s[i-1]) {
                        usarUnaMas = dp[i-1][j];
                    }
                    dp[i][j] = usarCero || usarUnaMas;
                } else {
                    dp[i][j] = false;
                }
            }
        }
        return dp[m][n];
    }
};

int main() {
    Solution sol;
    std::string s = "aa",
                p = "a";
    std::cout<<"Salida del problema 10: "<<sol.isMatch(s,p)<< std::endl;
}